### PR TITLE
Feature/refactor configuration reader and scheduledjob

### PR DESCRIPTION
**Pull request by andrewmarkham**
This includes
- a refactor of the configuration reader making the `ISynchronizerConfiguration` a dependency.  This allows for easier unit tests, and a way to deliver an alternate version.
- extended the `ScheduledJobDefinition`.  I added a new property _AutoRun_ which will run the scheduled job on environment change.
- similar fix to #2 

The new _AutoRun_ property is documented, and I have included some additional unit tests.

### CODE REVIEW

**Comment by andrewmarkham** [comment]
@ovelartelius Just you including in the PR as I didn't know how to assign a reviewer.

File: Addon.Episever.EnvironmentSynchronizer.Test/Configuration/ConfigurationReaderTests.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            Assert.AreEqual(true, data.ScheduledJobs[1].AutoRun);
+        }
+
+        [TestMethod()]
+        [DeploymentItem("test-configs\\no-settings.config")]
+        public void ReadConfiguration_SiteDefinition_No_Settings_Test()
+        {
+            // Arrange
+
+            ExeConfigurationFileMap configFileMap = new ExeConfigurationFileMap()
```
Comment (LinusEkstrom) [inline Addon.Episever.EnvironmentSynchronizer.Test/Configuration/ConfigurationReaderTests.cs]: Can use var here

File: Addon.Episever.EnvironmentSynchronizer.Test/Configuration/ConfigurationReaderTests.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            Assert.IsNotNull(data);
+        }
+
+        [TestMethod()]
+        [DeploymentItem("test-configs\\all-settings.config")]
+        public void ReadConfiguration_SiteDefinition_All_Settings_Test()
+        {
+            // Arrange
+
+            ExeConfigurationFileMap configFileMap = new ExeConfigurationFileMap()
```
Comment (LinusEkstrom) [inline Addon.Episever.EnvironmentSynchronizer.Test/Configuration/ConfigurationReaderTests.cs]: Can use var here

**Comment by ovelartelius** [comment]
@andrewmarkham Hi Andy. I apologize for the late reply. But I will look into ASAP.
/Regards
Ove

File: Addon.Episerver.EnvironmentSynchronizer/Configuration/ConfigurationReader.cs [hunk 3]
```diff
@@ -12,25 +12,31 @@ public class ConfigurationReader : IConfigurationReader
...
+			this.configuration = synchronizerConfiguration;
+        }
+
 		public SynchronizationData ReadConfiguration()
 		{
-			var config = new SynchronizerConfiguration();
 			var syncData = new SynchronizationData();
 
-			if(config.Settings == null)
+			if(this.configuration.Settings == null)
```
Comment (LinusEkstrom) [inline Addon.Episerver.EnvironmentSynchronizer/Configuration/ConfigurationReader.cs]: Remove usage of "this"

File: Addon.Episerver.EnvironmentSynchronizer/Configuration/ConfigurationReader.cs [hunk 4]
```diff
@@ -12,25 +12,31 @@ public class ConfigurationReader : IConfigurationReader
...
 
 			try
 			{
-				syncData.RunAsInitializationModule = config.Settings.RunAsInitializationModule;
-				syncData.RunInitializationModuleEveryStartup = config.Settings.RunInitializationModuleEveryStartup;
+				syncData.RunAsInitializationModule = this.configuration.Settings.RunAsInitializationModule;
+				syncData.RunInitializationModuleEveryStartup = this.configuration.Settings.RunInitializationModuleEveryStartup;
 
-				if (config.Settings.SiteDefinitions != null && config.Settings.SiteDefinitions.Count > 0)
+				if (this.configuration.Settings.SiteDefinitions != null && this.configuration.Settings.SiteDefinitions.Count > 0)
```
Comment (LinusEkstrom) [inline Addon.Episerver.EnvironmentSynchronizer/Configuration/ConfigurationReader.cs]: Remove usage of "this"

File: Addon.Episerver.EnvironmentSynchronizer/Configuration/ConfigurationReader.cs [hunk 5]
```diff
@@ -50,16 +56,17 @@ public SynchronizationData ReadConfiguration()
 					Logger.Information($"Found no site definitions to handle.");
 				}
 
-				if (config.Settings.ScheduleJobs != null && config.Settings.ScheduleJobs.Count > 0)
+				if (this.configuration.Settings.ScheduleJobs != null && this.configuration.Settings.ScheduleJobs.Count > 0)
```
Comment (LinusEkstrom) [inline Addon.Episerver.EnvironmentSynchronizer/Configuration/ConfigurationReader.cs]: Remove usage of "this"

**Review by LinusEkstrom — CHANGES_REQUESTED** [review]
Thanks for a good contribution, I just have some minor change requests before it's ready to merge.

File: Addon.Episerver.EnvironmentSynchronizer/Configuration/ConfigurationReader.cs (hunk 5 above)
Comment (andrewmarkham) [reply Addon.Episerver.EnvironmentSynchronizer/Configuration/ConfigurationReader.cs]: Removed :)

File: Addon.Episerver.EnvironmentSynchronizer/Configuration/ConfigurationReader.cs (hunk 4 above)
Comment (andrewmarkham) [reply Addon.Episerver.EnvironmentSynchronizer/Configuration/ConfigurationReader.cs]: Removed :)

File: Addon.Episerver.EnvironmentSynchronizer/Configuration/ConfigurationReader.cs (hunk 3 above)
Comment (andrewmarkham) [reply Addon.Episerver.EnvironmentSynchronizer/Configuration/ConfigurationReader.cs]: Removed :)

File: Addon.Episever.EnvironmentSynchronizer.Test/Configuration/ConfigurationReaderTests.cs (hunk 2 above)
Comment (andrewmarkham) [reply Addon.Episever.EnvironmentSynchronizer.Test/Configuration/ConfigurationReaderTests.cs]: Done

File: Addon.Episever.EnvironmentSynchronizer.Test/Configuration/ConfigurationReaderTests.cs (hunk 1 above)
Comment (andrewmarkham) [reply Addon.Episever.EnvironmentSynchronizer.Test/Configuration/ConfigurationReaderTests.cs]: Done

**Comment by andrewmarkham** [comment]
Hi @LinusEkstrom, 

I have amended as requested, and a few others which also had the same style.  As you can tell, I am a fan of '_this._' :).  Let me know if there is anything else.

Andy

File: Addon.Episerver.EnvironmentSynchronizer/Configuration/ConfigurationReader.cs [hunk 6]
```diff
@@ -12,25 +12,31 @@ public class ConfigurationReader : IConfigurationReader
 	{
 		private static readonly ILogger Logger = LogManager.GetLogger();
 
+		private readonly ISynchronizerConfiguration configuration;
```
Comment (LinusEkstrom) [inline Addon.Episerver.EnvironmentSynchronizer/Configuration/ConfigurationReader.cs]: private variables should be prefixed with _ aka _configuration. Sorry - missed this in the initial review.

**Review by LinusEkstrom — CHANGES_REQUESTED** [review]
Hi, quick changes, just noticed one more thing now with the this. gone.

File: Addon.Episerver.EnvironmentSynchronizer/Configuration/ConfigurationReader.cs (hunk 6 above)
Comment (andrewmarkham) [reply Addon.Episerver.EnvironmentSynchronizer/Configuration/ConfigurationReader.cs]: Hi, just updated.

**Comment by andrewmarkham** [comment]
Hi @ovelartelius / @LinusEkstrom 

Is there anything else that needs to be resolved before this PR is completed?

Thanks

**Comment by ovelartelius** [comment]
@andrewmarkham Hi. I have just compiled and added the newest dll v0.2.6.* in the master branch. I will now start working on make this as a nuget package. Thanks for helping!

**Comment by andrewmarkham** [comment]
> @andrewmarkham Hi. I have just compiled and added the newest dll v0.2.6.* in the master branch. I will now start working on make this as a nuget package. Thanks for helping!

Not a problem, nice to add something back.

**Pull request merged**